src/workerd/io/trace.c++

```diff
@@ -271,42 +271,32 @@ Trace::Exception::Exception(rpc::Trace::Exception::Reader reader)
       name(kj::str(reader.getName())),
       message(kj::str(reader.getMessage())) {}
 
-Tracer::Span::Span(kj::Own<Tracer> tracer, kj::Maybe<Jaeger::SpanData> spanData,
-    kj::TimePoint durationStartTime)
-    : tracer(kj::mv(tracer)), spanData(kj::mv(spanData)),
-      durationStartTime(durationStartTime) {}
-
-Tracer::Span::~Span() noexcept(false) {
-  KJ_IF_MAYBE(d, spanData) {
-    d->duration = isPredictableModeForTest() ? kj::Duration() :
-        (kj::systemPreciseMonotonicClock().now() - durationStartTime);
-    if (droppedLogs > 0) {
-      d->logs.add(Jaeger::SpanData::Log {
-        .timestamp = d->startTime + d->duration,
-        .tag = {
-          .key = "dropped_logs"_kj,
-          .value = int64_t(droppedLogs),
-        }
-      });
-    }
-    auto& submitter = KJ_REQUIRE_NONNULL(tracer->jaegerSpanSubmitter);
-    submitter.submitSpan(kj::mv(*d));
-  }
+SpanBuilder& SpanBuilder::operator=(SpanBuilder &&other) {
+  end();
+  state = kj::mv(other.state);
+  return *this;
+}
+
+SpanBuilder::~SpanBuilder() noexcept(false) {
+  end();
 }
 
-kj::Maybe<Jaeger::SpanContext> Tracer::Span::getSpanContext() {
-  return spanData.map([](Jaeger::SpanData& d) { return d.context; });
+void SpanBuilder::end() {
+  KJ_IF_MAYBE(s, state) {
```

Comment:
This won't double report if we call end multiple times right? I'm not seeing anything that sets the state to null or anything to prevent it...

Reply:
Another good catch, this is indeed a bug. Fixed, and squashed commits since they were kinda fixup-y.